File: Heaps/lc264.py
'''
264. Ugly Number II
Solved
Medium
Topics
Companies
Hint
An ugly number is a positive integer whose prime factors are limited to 2, 3, and 5.

Given an integer n, return the nth ugly number.



Example 1:

Input: n = 10
Output: 12
Explanation: [1, 2, 3, 4, 5, 6, 8, 9, 10, 12] is the sequence of the first 10 ugly numbers.
Example 2:

Input: n = 1
Output: 1
Explanation: 1 has no prime factors, therefore all of its prime factors are limited to 2, 3, and 5.


Constraints:

1 <= n <= 1690
'''
import heapq


class Solution:
    def nthUglyNumber(self, n: int) -> int:
        heap = []
        current_ugly_number = 1
        ugly_number_count = 1
        factors = [1, 2, 3]
        ugly_number_set = {1}

        while ugly_number_count < n:
            for factor in factors:
                ugly_number = factor * current_ugly_number
                if ugly_number not in ugly_number_set:
                    heapq.heappush(heap, ugly_number)
                    ugly_number_set.add(ugly_number)
            current_ugly_number = heapq.heappop(heap)
            ugly_number_count += 1

        return current_ugly_number

File: Heaps/lc264.java
//264. Ugly Number II//        Solved//        Medium//        Topics//        Companies//        Hint//        An ugly number is a positive integer whose prime factors are limited to 2, 3, and 5.////        Given an integer n, return the nth ugly number.////////        Example 1:////        Input: n = 10//        Output: 12//        Explanation: [1, 2, 3, 4, 5, 6, 8, 9, 10, 12] is the sequence of the first 10 ugly numbers.//        Example 2:////        Input: n = 1//        Output: 1//        Explanation: 1 has no prime factors, therefore all of its prime factors are limited to 2, 3, and 5.//////        Constraints:////        1 <= n <= 1690package Heaps;import java.util.HashSet;import java.util.PriorityQueue;import java.util.Queue;import java.util.Set;public class lc264 {    class Solution {        public int nthUglyNumber(int n) {            Queue<Long> heap = new PriorityQueue<>();            Set<Long> uglyNumberSet = new HashSet<>();            Long currentUglyNumber = 1L;            int[] factors = {2, 3, 5};            int uglyNumberCount = 1;            while (uglyNumberCount < n) {                for (int factor: factors) {                    Long uglyNumber = factor * currentUglyNumber;                    if (!uglyNumberSet.contains(uglyNumber)) {                        uglyNumberSet.add(uglyNumber);                        heap.add(uglyNumber);                    }                }                currentUglyNumber = heap.remove();                uglyNumberCount += 1;            }            return Math.toIntExact(currentUglyNumber);        }    }}